slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala

```diff
@@ -356,6 +356,22 @@ trait JdbcActionComponent extends SqlActionComponent { self: JdbcProfile =>
 
   //////////////////////////////////////////////////////////// InsertActionComposer Traits
 
+
+  /**
+   * Specifies how multi-row operations are performed
+   */
+  sealed trait StatementOption
+
+  object StatementOption {
+
+    /** Multi-row operations are executed in a single statement. */
+    case object SingleStatement extends StatementOption
+
+    /** Multi-row operations are executed with batch of commands. */
+    case object Batch extends StatementOption
+
+  }
```

Comment:
I thought that there are not always only two options for optimization, so I decided to specify an enum instead of a `Boolean` type.

Reply:
Great, that's better than a boolean. But unless this has applicability for more statements than just insert, I would make its name more specific. So maybe `InsertAllMode` or something.

Also, the naming is inconsistent -- why not `BatchStatement` or `Single`? I would omit the `Statement` suffix since it would be obvious because you're dotting into it (`StatementOption.Single`).

Also, since batch might not be supported perhaps the second option is misleading?

Actually since "MultiInsert" is a term that's used already, how about `MultiInsertStrategy.EachRow` and `.AllRows`?

I can make this change on Sunday if you like it and don't get to it first.

Another thought, since it's just a data type we could have another option:

```scala
    case class ChunkedRows(size: Int) extends MultiInsertStrategy
```

which would mean to do `size` rows per statement, so it would use multi-row VALUES clauses but if the number of rows to insert exceeds `size` it will still generate multiple statements. You mentioned the tradeoff, that doing them all in a single statement could cause issues sometimes. Having this option could allow optimizing the tradeoff.

We could implement this by giving `MultiInsertStrategy` a `def size: Int` and `EachRow` would override it to `1` and `AllRows` (or whatever we name them) would override it to `Int.MaxValue`. Then the implementation wouldn't have to pattern match. It could just read `size`.

That said, at that point they are isomorphic to their `size` value. Which means potentially instead of introducing a new data type and using it as the parameter, we could simply have a `rowsPerStatement: Int = Int.MaxValue` parameter.

Reply:
> So maybe InsertAllMode or something.

I was wondering if I need to take care of insertOrUpdate, but I am not particular about it.

Reply:
> Also, the naming is inconsistent -- why not BatchStatement or Single? I would omit the Statement suffix since it would be obvious because you're dotting into it (StatementOption.Single).

Batch/Single looks good. (To be precise, Batch is the use of JDBC functions, and Single is the SQL syntax.)

Reply:
> > So maybe InsertAllMode or something.
> 
> I was wondering if I need to take care of insertOrUpdate, but I am not particular about it.

Maybe a better naming scheme is `RowsPerStatement.{One, All}` (meaning the data type would be called `RowsPerStatement` and the choices would be `One` and `All`)

Reply:
> Actually since "MultiInsert" is a term that's used already, how about MultiInsertStrategy.EachRow and .AllRows?

My hope is that the API users can guess whether they are using JDBC functionality (= batch) or SQL statement. It seems a little too abstract in that regard.

Reply:
> ChunkedRows

It would be nice to use such an interface as a first improvement.
However, after looking at it a bit, there are several options for future support of Oracle(https://database.guide/4-ways-to-insert-multiple-rows-in-oracle/), and this may be the case with other profiles. Therefore, I feel it is better to use enum as an interface than a simple Int type, even if it is only `ChunkedRows` at first.